rivertest/worker.go

```diff
@@ -70,86 +74,129 @@ func NewWorker[T river.JobArgs, TTx any](tb testing.TB, driver riverdriver.Drive
+	tb.Helper()
+
+	result, err := w.client.InsertTx(ctx, tx, args, opts)
+	if err != nil {
+		tb.Fatalf("failed to insert job: %s", err)
+	}
+	return result.Job
+}
+
 // Work allocates a synthetic job with the provided arguments and optional
 // insert options, then works it.
-func (w *Worker[T, TTx]) Work(ctx context.Context, tb testing.TB, args T, opts *river.InsertOpts) error {
+func (w *Worker[T, TTx]) Work(ctx context.Context, tb testing.TB, tx TTx, args T, opts *river.InsertOpts) error {
 	tb.Helper()
-	job := makeJobFromArgs(tb, w.config, args, opts)
-	return w.WorkJob(ctx, tb, job)
+
+	job := w.insertJob(ctx, tb, tx, args, opts)
+	return w.WorkJob(ctx, tb, tx, job)
 }
 
 // WorkTx allocates a synthetic job with the provided arguments and optional
 // insert options, then works it in the given transaction.
 func (w *Worker[T, TTx]) WorkTx(ctx context.Context, tb testing.TB, tx TTx, args T, opts *river.InsertOpts) error {
 	tb.Helper()
-	job := makeJobFromArgs(tb, w.config, args, opts)
+
+	job := w.insertJob(ctx, tb, tx, args, opts)
 	return w.WorkJobTx(ctx, tb, tx, job)
 }
 
 // WorkJob works the provided job. The job must be constructed to be a realistic
 // job using external logic prior to calling this method. Unlike the other
 // variants, this method offers total control over the job's attributes.
-func (w *Worker[T, TTx]) WorkJob(ctx context.Context, tb testing.TB, job *river.Job[T]) error {
+func (w *Worker[T, TTx]) WorkJob(ctx context.Context, tb testing.TB, tx TTx, job *rivertype.JobRow) error {
 	tb.Helper()
-
-	var exec riverdriver.Executor
-	if w.client.Driver().HasPool() {
-		exec = w.client.Driver().GetExecutor()
-	}
-
-	return w.workJobExec(ctx, tb, exec, job)
+	return w.workJob(ctx, tb, tx, job)
 }
 
 // WorkJobTx works the provided job in the given transaction. The job must be
 // constructed to be a realistic job using external logic prior to calling this
 // method. Unlike the other variants, this method offers total control over the
 // job's attributes.
-func (w *Worker[T, TTx]) WorkJobTx(ctx context.Context, tb testing.TB, tx TTx, job *river.Job[T]) error {
+func (w *Worker[T, TTx]) WorkJobTx(ctx context.Context, tb testing.TB, tx TTx, job *rivertype.JobRow) error {
 	tb.Helper()
-	return w.workJobExec(ctx, tb, w.client.Driver().UnwrapExecutor(tx), job)
+	return w.workJob(ctx, tb, tx, job)
 }
 
-func (w *Worker[T, TTx]) workJobExec(ctx context.Context, tb testing.TB, exec riverdriver.Executor, job *river.Job[T]) error {
+func (w *Worker[T, TTx]) workJob(ctx context.Context, tb testing.TB, tx TTx, job *rivertype.JobRow) error {
 	tb.Helper()
 
+	exec := w.client.Driver().UnwrapExecutor(tx)
+	subscribeCh := make(chan []jobcompleter.CompleterJobUpdated, 1)
+	archetype := riversharedtest.BaseServiceArchetype(tb)
+	completer := jobcompleter.NewInlineCompleter(archetype, exec, w.client.Pilot(), subscribeCh)
```

Comment:
Would be a very minor saving, but I wonder if we should let this tolerate a `nil` subscribe channel so we can skip the allocation and the frontmatter code.

Reply:
I actually ended up leveraging this in the updated version in order to get the post-completion job row for the new return struct. We might consider refactoring it as there's a handful of channels at play, lmk once you get to review the new version (coming shortly).